### PR TITLE
Feat(Compiler-Base): Add code context pendant.

### PR DESCRIPTION
**Pull request by zong-zhe**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

issue #115

#### 2. What is the scope of this PR (e.g. component or file name):

compiler_base/error/src/components.rs

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

add errors for `Compiler-Base-Error` by `anyhow`.
defines error types for `Compiler-Base-Error` in `compiler_base/error/errors.rs`.
add `CodeSnippet` in `compiler_base/error/src/diagnostic/components.rs` to show code context in diagnostic.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

Add unit tests in `compiler_base/error/src/tests.rs`.

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/compiler_base/error/Cargo.toml [hunk 1]
```diff
@@ -14,4 +14,4 @@ unic-langid = {version="0.9.0", features = ["macros"]}
 fluent = "0.16.0"
 termcolor = "1.0"
 walkdir = "2"
-anyhow = "1.0"
\ No newline at end of file
+anyhow = "1.0"
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/Cargo.toml:17]: Add newline at end of file.

[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/Cargo.toml:17]: fixed

File: kclvm/compiler_base/error/src/diagnostic/test_datas/main.k [hunk 2]
```diff
@@ -0,0 +1,8 @@
+schema Person:
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/test_datas/main.k]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 3]
```diff
@@ -52,3 +57,389 @@ impl Component<DiagnosticStyle> for Label {
...
+    /// ```
+    pub fn new_with_default_label(
+        start: usize,
+        end: usize,
+        style: Option<DiagnosticStyle>,
+    ) -> Self {
+        Self {
+            start,
+            end,
+            label: "^".to_string(),
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/span/src/lib.rs [hunk 4]
```diff
@@ -9,9 +9,13 @@
...
 pub use rustc_span::fatal_error;
-pub use span::{BytePos, Span, DUMMY_SP};
+pub use span::{BytePos, Span, SpanData, DUMMY_SP};
 
 pub type SourceMap = rustc_span::SourceMap;
 pub type SourceFile = rustc_span::SourceFile;
 pub type FilePathMapping = rustc_span::source_map::FilePathMapping;
 pub type Loc = rustc_span::Loc;
+
+pub fn span_to_filename_string(span: &Span, sm: &SourceMap) -> String {
```
Comment (Peefy) [inline kclvm/compiler_base/span/src/lib.rs:19]: more documents on the fuction and make it `#[inline]`?

File: kclvm/compiler_base/span/src/span.rs [hunk 5]
```diff
@@ -2,4 +2,9 @@ use rustc_span;
 
 pub type BytePos = rustc_span::BytePos;
 pub type Span = rustc_span::Span;
+pub type SpanData = rustc_span::SpanData;
 pub const DUMMY_SP: Span = rustc_span::DUMMY_SP;
+
+pub fn new_byte_pos(arg: u32) -> rustc_span::BytePos {
```
Comment (Peefy) [inline kclvm/compiler_base/span/src/span.rs:8]: more documents on the fuction and make it #[inline]?

Comment (zong-zhe) [reply kclvm/compiler_base/span/src/span.rs:17]: fixed

File: kclvm/compiler_base/span/src/lib.rs (hunk 4 above)
Comment (zong-zhe) [reply kclvm/compiler_base/span/src/lib.rs:45]: fixed

**Comment by ldxdl** [comment]
This PR includes a lot of code changes, which can be split according to the purpose

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 6]
```diff
@@ -52,3 +57,394 @@ impl Component<DiagnosticStyle> for Label {
...
+///
+/// ```ignore
+/// int test = 0;
+///     ^^^^ This is an underline under variable `test`
+/// ```
+pub struct UnderLine {
+    start: usize,
+    end: usize,
+    label: String,
+    style: Option<DiagnosticStyle>,
```
Comment (ldxdl) [inline kclvm/compiler_base/error/src/diagnostic/components.rs:71]: I prefer that the style is decoupled with each component itself
style take effect to all components and is not a part of any component

**Pull request closed without merging**